Don't cast `VersionConstraint` to `MutableVersionConstraint` in `DefaultMutableMinimalDependency`

Similar to: https://github.com/gradle/gradle/issues/22557

When implementing a custom `VersionCatalog` you may need a custom implementation of `VersionConstraint`
`VersionConstraint` is not marked with `@HasInternalProtocol` 

https://github.com/gradle/gradle/blob/68d50c494d2e85a75128dcc0fe7203018807dc84/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMutableMinimalDependency.java#L24-L27

But any use of a custom implementation of 'VersionConstraint' leads to this error. 

```
Caused by: org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type org.gradle.api.internal.artifacts.dependencies.DefaultMutableMinimalDependency.
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:69)
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:55)
        at org.gradle.api.internal.notations.DependencyNotationParser$MinimalExternalDependencyNotationConverter.convert(DependencyNotationParser.java:152)
        at org.gradle.api.internal.notations.DependencyNotationParser$MinimalExternalDependencyNotationConverter.convert(DependencyNotationParser.java:139)
        at org.gradle.internal.typeconversion.TypeFilteringNotationConverter.convert(TypeFilteringNotationConverter.java:33)
        at org.gradle.internal.typeconversion.CompositeNotationConverter.convert(CompositeNotationConverter.java:34)
        at org.gradle.internal.typeconversion.NotationConverterToNotationParserAdapter.parseNotation(NotationConverterToNotationParserAdapter.java:31)
        at org.gradle.internal.typeconversion.ErrorHandlingNotationParser.parseNotation(ErrorHandlingNotationParser.java:48)
        at org.gradle.api.internal.artifacts.DefaultDependencyFactory.createDependency(DefaultDependencyFactory.java:79)
        at org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.create(DefaultDependencyHandler.java:169)
        at org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.lambda$mapDependencyProvider$0(DefaultDependencyHandler.java:222)
        at org.gradle.api.internal.provider.ValueSupplier$Present.transform(ValueSupplier.java:541)
        at org.gradle.api.internal.provider.TransformBackedProvider.mapValue(TransformBackedProvider.java:82)
        at org.gradle.api.internal.provider.TransformBackedProvider.calculateOwnValue(TransformBackedProvider.java:74)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateOwnPresentValue(AbstractMinimalProvider.java:72)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:92)
        at org.gradle.api.internal.DefaultDomainObjectCollection.addLater(DefaultDomainObjectCollection.java:286)
        at org.gradle.api.internal.DelegatingDomainObjectSet.addLater(DelegatingDomainObjectSet.java:100)
        at org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.doAddProvider(DefaultDependencyHandler.java:212)
        at org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.doAdd(DefaultDependencyHandler.java:187)
        at org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.add(DefaultDependencyHandler.java:121)
        at org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.add(DefaultDependencyHandler.java:115)
        at org.gradle.kotlin.dsl.support.delegates.DependencyHandlerDelegate.add(DependencyHandlerDelegate.kt:55)
        at gradle.kotlin.dsl.accessors._5b841d749b44c4d634f7cfea3ed45134.ImplementationConfigurationAccessorsKt.implementation(ImplementationConfigurationAccessors.kt:71)
        ... 
Caused by: java.lang.IllegalArgumentException: Unable to determine constructor argument #2: value MyVersionConstraint@12fa2b93 not assignable to type MutableVersionConstraint.
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.verifyParameters(DependencyInjectingInstantiator.java:146)
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.convertParameters(DependencyInjectingInstantiator.java:119)
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:62)
        ... 215 more

--- a/subprojects/core-api/src/main/java/org/gradle/api/artifacts/VersionConstraint.java
+++ b/subprojects/core-api/src/main/java/org/gradle/api/artifacts/VersionConstraint.java
@@ -16,6 +16,7 @@
 package org.gradle.api.artifacts;
 
 import org.gradle.api.Describable;
+import org.gradle.internal.HasInternalProtocol;
 import org.gradle.internal.scan.UsedByScanPlugin;
 
 import javax.annotation.Nullable;
@@ -92,6 +93,7 @@ import java.util.List;
  * @since 4.4
  */
 @UsedByScanPlugin
+@HasInternalProtocol
 public interface VersionConstraint extends Describable {
     /**
      * The branch to select versions from. When not {@code null} selects only those versions that were built from the specified branch.

